File: 00-每日一题/20200321_51.cpp
//
// Created by zanbo on 2020/3/21.
//

class Solution {
public:
    vector<vector<string>> res;
    vector<vector<string>> solveNQueens(int n) {
        vector<string> temp(n,string(n,'.'));
        dfs(0,temp,n);
        return res;
    }

    void dfs(int row, vector<string>& temp, int n)
    {
        if(row==n)
        {
            res.push_back(temp);
            return;
        }
        for(int j=0;j<n;j++)
        {
            if(isvalid(row,j,temp,n))
            {
                temp[row][j] = 'Q';
                dfs(row+1,temp,n);
                temp[row][j] = '.';
            }
        }
    }

    bool isvalid(int i,int j,vector<string> temp,int n)
    {
        // 分别判断所在列和两个斜线 是否有Q
        for(int k=0;k<i;k++)
        {
            if(temp[k][j]=='Q')
                return false;
        }
        for(int k=0;i-k>=0 && j-k>=0;k++)
        {
            if(temp[i-k][j-k]=='Q')
                return false;
        }
        for(int k=0;i-k>=0 && j+k<n;k++)
        {
            if(temp[i-k][j+k]=='Q')
                return false;
        }
        return true;
    }

};

File: 00-每日一题/20200321_51.py
class Solution(object):
    col = None
    dia = None
    dia_back = None

    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """

        res = []
        if n == 0:
            return []

        self.col = [False for _ in range(n)]
        # 对角线
        # 正对角线 \
        # 两数坐标之和从0到2n-2
        self.dia = [False for _ in range(2 * n - 1)]
        # 反对角线 /
        # 两数坐标只差由-n+1到n-1
        # 加上n-1后，可定位至0到2n-2
        self.dia_back = [False for _ in range(2 * n - 1)]
        self._dfs(n, 0, [], res)
        return res


    def draw_res(self, lists):
        res = []
        for q in lists:
            line = ''
            for i in range(len(lists)):
                if i != q:
                    line += '.'
                else:
                    line += 'Q'
            res.append(line)

        return res

    def _dfs(self, n, row, pre, res):
        # 回溯结束
        if row == n:
            res.append(self.draw_res(pre))
            return

        # 遍历
        for i in range(n):
            if self.col[i] is False and self.dia[i + row] is False \
                    and self.dia_back[i - row + n - 1] is False:
                # 状态设置
                self.col[i] = True
                self.dia[i + row] = True
                self.dia_back[i - row + n - 1] = True
                pre.append(i)

                self._dfs(n, row+1, pre, res)

                # 回溯，清除状态
                self.col[i] = False
                self.dia[i + row] = False
                self.dia_back[i - row + n - 1] = False
                pre.pop()

if __name__ == '__main__':
    obj = Solution()
    while True:
        n = int(input())
        res = obj.solveNQueens(n)
        print(res)
